lib/auth/access.go

```diff
@@ -144,3 +148,71 @@ func (a *Server) DeleteLock(ctx context.Context, lockName string) error {
 	}
 	return nil
 }
+
+// checkRoleRulesConstraint checks if the request will result in having
+// no roles with rules to upsert roles.
+func (s *Server) checkRoleRulesConstraint(ctx context.Context, targetRole types.Role, request string) error {
+	rolesWithUpdateRolesRule, err := s.getRolesWithUpdateRolesRule(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	isRoleLoosingUpdateRolesRule := Contains(rolesWithUpdateRolesRule, targetRole.GetName()) && !roleHasUpdateRolesRule(targetRole)
+
+	if !isRoleLoosingUpdateRolesRule {
+		return nil
+	}
+
+	allUsers, err := s.Identity.GetUsers(false)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	localUsersWithPermissionToEditRoles := Filter(allUsers, func(u types.User) bool {
+		isLocalUser := u.GetCreatedBy().Connector == nil
+
+		rolesWithoutTarget := Filter(u.GetRoles(), func(role string) bool {
+			return role != targetRole.GetName()
+		})
+
+		return isLocalUser && Some(rolesWithoutTarget, func(role string) bool {
+			return Contains(rolesWithUpdateRolesRule, role)
+		})
+	})
+
+	if len(localUsersWithPermissionToEditRoles) > 0 {
+		return nil
+	}
+
+	log.Warnf("Failed to %s update role. This operation will cause no user to be able to edit roles.", request)
+	return trace.BadParameter("Failed to %s update role. This operation will cause no user to be able to edit roles.", request)
+}
+
+// returns a list of roles that have a update rule associated to the role resource.
+func (a *Server) getRolesWithUpdateRolesRule(ctx context.Context) ([]string, error) {
+	allRoles, err := a.Access.GetRoles(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	getRolesWithUpdateRolesRule := Filter(allRoles, roleHasUpdateRolesRule)
+
+	return Map(getRolesWithUpdateRolesRule, func(r types.Role) string {
+		return r.GetName()
+	}), nil
+}
+
+// checks if role has permission to edit roles
+func roleHasUpdateRolesRule(role types.Role) bool {
+	return Some(role.GetRules(types.Allow), func(rule types.Rule) bool {
+		if !rule.HasResource(types.KindRole) {
+			return false
+		}
+
+		if !rule.HasVerb(types.VerbUpdate) {
+			return false
+		}
+
+		return true
+	})
```

Comment:
The FP is fun, but I think boring `for` loops are the idiomatic way to do this in Go (and they're somewhat more efficient)

Reply:
I have removed fp module: https://github.com/gravitational/teleport/commit/d4e2fb2c21094c1de433e01633a104cc616e9467